### PR TITLE
Add routing to analyses nodes

### PR DESCRIPTION
**Pull request by rubenmoya**
## Context
Before we only had the analyses nodes linked in here:

screen shot 2018-01-10 at 12 42 56

This PR adds that link in the analysis form too:

screen shot 2018-01-10 at 12 42 45

## Changes
- Added deep-linking to analysis nodes
- Replaced the model used to manage `selectedNodeId` with routing

## Acceptance
When there is only one analysis and it's not saved:
- [ ] The url has the correct node id
- [ ] Going back renders the placeholder and removes the node id from the url
- [ ] Refresh renders the placeholders and removes the node id from the route
- [ ] Going back and then forward does nothing 🤷‍♂️

With all nodes saved:
- [ ] The url has the correct node id
- [ ] The url changes when moving between nodes
- [ ] Going back selects the correct node
- [ ] Going forward selects the correct node
- [ ] Refresh selects the correct node

With multiple analysis applied and the selected one is not applied:
- [ ] The url has the correct node id
- [ ] The url changes when moving between nodes
- [ ] Going back selects the previous analysis
- [ ] Going forward (after clicking in another one) don't delete and selects the unsaved analysis
- [ ] Refresh deletes the unsaved analysis and selects the last one applied

### CODE REVIEW

**Comment by jesusbotella** [comment]
I have discovered that when you add a new analysis and click on the previous one on the list, it sets both ones as selected, instead of deselecting the previous one. Rubén just move it back to In Progress!

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-view.js [hunk 1]
```diff
@@ -282,7 +284,14 @@ module.exports = CoreView.extend({
 
           analysisFormsCollection.addHead(analysisPayload);
         } else {
-          selectedNodeId = self._layerDefinitionModel.get('source');
+          var lastAnalysisNode = analysisFormsCollection.at(0);
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-view.js]: this needs tests coverage

File: lib/assets/core/javascripts/cartodb3/editor.js [hunk 2]
```diff
@@ -372,13 +372,25 @@ var settingsCollection = new Backbone.Collection(SettingsOptions(overlaysCollect
 
 Router.init();
 
+var handleAnalysesRoute = function (routeModel) {
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor.js]: don't we have a suite for this?

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-content-view.js [hunk 3]
```diff
@@ -185,16 +183,18 @@ module.exports = CoreView.extend({
...
 
-  _getSelectedNodeId: function () {
-    return this.model.get('selectedNodeId');
-  },
-
   _onLayerDefSourceChanged: function () {
     this._analysisFormsCollection.removeUselessModels();
   },
 
+  _onAnalysisFormsCollectionSync: function (model) {
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-content-view.js]: tests

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-service.js [hunk 4]
```diff
@@ -71,6 +74,18 @@ module.exports = (function () {
           this._stackLayoutView.model.goToStep(1, layerDefinitionModel, 'layer-content', 'analyses', analysisFormAttrs);
         }
       }, this);
+    },
+
+    saveNotAppliedAnalysis: function (analysis) {
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-service.js]: tests

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-workflow-list-view.js [hunk 5]
```diff
@@ -8,18 +8,13 @@ var utils = require('../../../../helpers/utils');
...
-  'layerId'
+  'layerId',
+  'selectedNodeId'
 ];
 
 module.exports = CoreView.extend({
   initialize: function (opts) {
     checkAndBuildOpts(opts, REQUIRED_OPTS, this);
-
-    if (!opts.model) {
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-workflow-list-view.js]: model should go in `REQUIRED_OPTS` or, if we don't want `this._model` leave this as is

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-workflow-view.js [hunk 6]
```diff
@@ -1,23 +1,20 @@
...
+  'layerDefinitionModel',
+  'selectedNodeId'
 ];
 
 module.exports = CoreView.extend({
 
   initialize: function (opts) {
     checkAndBuildOpts(opts, REQUIRED_OPTS, this);
 
-    if (!opts.model) {
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-workflow-view.js]: model should go in `REQUIRED_OPTS` or, if we don't want `this._model` leave this as is

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-forms-collection.js [hunk 7]
```diff
@@ -33,19 +34,29 @@ module.exports = Backbone.Collection.extend({
...
+        var attrs = _.extend({ persisted: true }, nodeDefModel.attributes);
         attrsList.push(attrs);
       });
+
       this.reset(attrsList);
     }
   },
 
   addHead: function (attrs) {
+    AnalysesService.saveNotAppliedAnalysis(attrs);
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analysis-forms-collection.js]: tests

File: lib/assets/core/javascripts/cartodb3/routes/router.js [hunk 8]
```diff
@@ -107,12 +107,14 @@ var Router = (function () {
       this.navigate(populateRoute(ROUTE_LAYER_DATA_TAB, { layerId: layerId }));
     },
 
-    goToAnalysisTab: function (layerId) {
-      this.navigate(populateRoute(ROUTE_LAYER_ANALYSES_TAB, { layerId: layerId }));
+    goToAnalysisTab: function (layerId, options) {
```
Comment (matallo) [inline lib/assets/core/javascripts/cartodb3/routes/router.js]: tests

File: lib/assets/core/javascripts/cartodb3/editor.js (hunk 2 above)
Comment (rjimenezda) [reply lib/assets/core/javascripts/cartodb3/editor.js]: there are no specs for editor afaik

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-service.js (hunk 4 above)
Comment (rubenmoya) [reply lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-service.js]: ✔️  Done

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-content-view.js (hunk 3 above)
Comment (rubenmoya) [reply lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-views/analyses/analyses-content-view.js]: ✔️  Done

File: lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-view.js (hunk 1 above)
Comment (rubenmoya) [reply lib/assets/core/javascripts/cartodb3/editor/layers/layer-content-view.js]: ✔️  Done

File: lib/assets/core/javascripts/cartodb3/routes/router.js (hunk 8 above)
Comment (rubenmoya) [reply lib/assets/core/javascripts/cartodb3/routes/router.js]: we don't have tests for the router yet :(

Comment (matallo) [reply lib/assets/core/javascripts/cartodb3/routes/router.js]: YET

Comment (matallo) [reply lib/assets/core/javascripts/cartodb3/routes/router.js]: now we do

Comment (rubenmoya) [reply lib/assets/core/javascripts/cartodb3/routes/router.js]: VAMOOOOOOS! 💪

**Pull request merged**